lucene/core/src/java/org/apache/lucene/search/FuzzyTermsEnum.java

```diff
@@ -92,76 +105,62 @@
-      throw new IllegalArgumentException("max edits must be 0.." + LevenshteinAutomata.MAXIMUM_SUPPORTED_DISTANCE + ", inclusive; got: " + maxEdits);
-    }
-    if (prefixLength < 0) {
-      throw new IllegalArgumentException("prefixLength cannot be less than 0");
-    }
+  public FuzzyTermsEnum(Terms terms, AttributeSource atts, Term term, int termLength,
+      final int maxEdits, CompiledAutomaton[] automata) throws IOException {
+
     this.maxEdits = maxEdits;
     this.terms = terms;
     this.term = term;
-    
-    // convert the string into a utf32 int[] representation for fast comparisons
-    this.termText = stringToUTF32(term.text());
-    this.termLength = termText.length;
+    this.atts = atts;
+    this.termLength = termLength;
 
-    this.dfaAtt = atts.addAttribute(LevenshteinAutomataAttribute.class);
     this.maxBoostAtt = atts.addAttribute(MaxNonCompetitiveBoostAttribute.class);
+    this.boostAtt = atts.addAttribute(BoostAttribute.class);
 
-    // NOTE: boostAtt must pulled from attributes() not from atts!  This is because TopTermsRewrite looks for boostAtt from this TermsEnum's
-    // private attributes() and not the global atts passed to us from MultiTermQuery:
-    this.boostAtt = attributes().addAttribute(BoostAttribute.class);
-
-    //The prefix could be longer than the word.
-    //It's kind of silly though.  It means we must match the entire word.
-    this.realPrefixLength = prefixLength > termLength ? termLength : prefixLength;
-    this.transpositions = transpositions;
-
-    CompiledAutomaton[] prevAutomata = dfaAtt.automata();
-    if (prevAutomata == null) {
-      prevAutomata = new CompiledAutomaton[maxEdits+1];
-      Automaton[] automata = buildAutomata(termText, prefixLength, transpositions, maxEdits);
-      for (int i = 0; i <= maxEdits; i++) {
-        try {
-          prevAutomata[i] = new CompiledAutomaton(automata[i], true, false);
-        } catch (TooComplexToDeterminizeException e) {
-          throw new FuzzyTermsException(term.text(), e);
-        }
-      }
-      // first segment computes the automata, and we share with subsequent segments via this Attribute:
-      dfaAtt.setAutomata(prevAutomata);
-    }
+    this.automata = automata;
 
-    this.automata = prevAutomata;
     bottom = maxBoostAtt.getMaxNonCompetitiveBoost();
     bottomTerm = maxBoostAtt.getCompetitiveTerm();
     bottomChanged(null);
   }
 
   /**
-   * Builds a binary Automaton to match a fuzzy term
-   * @param text            the term to match
-   * @param prefixLength    length of a required common prefix
-   * @param transpositions  {@code true} if transpositions should count as a single edit
-   * @param maxEdits        the maximum edit distance of matching terms
+   * Sets the maximum non-competitive boost, which may allow switching to a
+   * lower max-edit automaton at run time
+   */
+  public void setMaxNonCompetitiveBoost(float boost) {
+    this.maxBoostAtt.setMaxNonCompetitiveBoost(boost);
```

Comment:
Does this need to call `bottomChanged`?

Reply:
Thinking on this further... would it make sense to somehow check the terms before building all of the automata, and then possibly skip building the ed=2 case to begin with? Maybe checking the terms is more expensive than building the automaton, so we only care about saving time on how much we use the automaton at runtime? I'm not sure here and haven't done the profiling necessary to be confident.

Reply:
> Does this need to call bottomChanged?

No, that's an internal call only, and it's delayed anyway so as to preserve term stats after calling `next()`.

> Maybe checking the terms is more expensive than building the automaton

I'd be surprised, unless the input term for the automaton is very long, but it's probably worth profiling and checking this in a followup.